loadtest/main.go

```diff
@@ -56,49 +70,112 @@ func init() {
 	client := NewLoadTestClient(config)
 	client.SetValidators()
 
-	if config.TxsPerBlock < config.MsgsPerTx {
-		panic("Must have more TxsPerBlock than MsgsPerTx")
-	}
-
 	configString, _ := json.Marshal(config)
 	fmt.Printf("Running with \n %s \n", string(configString))
 
-	fmt.Printf("%s - Starting block prepare\n", time.Now().Format("2006-01-02T15:04:05"))
-	workgroups, sendersList := client.BuildTxs()
+	txQueue := make(chan []byte, 10000)
+	done := make(chan struct{})
+	numProducers := 5
+	var wg sync.WaitGroup
 
-	go client.SendTxs(workgroups, sendersList)
+	// Catch OS signals for graceful shutdown
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
 
-	// Waits until SendTx is done processing before proceeding to write and validate TXs
-	client.GatherTxHashes()
+	numTxsPerProducerPerSecond := 1
+	if int(config.TargetTps) > numProducers {
+		numTxsPerProducerPerSecond = int(config.TargetTps) / numProducers
+	}
 
-	// Records the resulting TxHash to file
-	client.WriteTxHashToFile()
-	fmt.Printf("%s - Finished\n", time.Now().Format("2006-01-02T15:04:05"))
+	ticker := time.NewTicker(5 * time.Second)
+	start := time.Now()
+	var producedCount int64 = 0
+	var sentCount int64 = 0
+	var prevSentCount int64 = 0
+	var blockHeights []int
+	var blockTimes []string
+	var startHeight = getLastHeight(config.BlockchainEndpoint)
+	fmt.Printf("Starting loadtest producers\n")
+	for i := 0; i < numProducers; i++ {
+		wg.Add(1)
+		go client.BuildTxs(txQueue, i, numTxsPerProducerPerSecond, &wg, done, &producedCount)
+	}
 
-	// Validate Tx will close the connection when it's done
-	go client.ValidateTxs()
+	fmt.Printf("Starting loadtest consumers\n")
+	go client.SendTxs(txQueue, done, &sentCount)
+	// Statistics reporting goroutine
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				currHeight := getLastHeight(config.BlockchainEndpoint)
+
+				for i := startHeight; i <= currHeight; i++ {
+					_, blockTime, err := getTxBlockInfo(config.BlockchainEndpoint, strconv.Itoa(i))
+					if err != nil {
+						fmt.Printf("Encountered error scraping data: %s\n", err)
+						return
+					}
+					blockHeights = append(blockHeights, i)
+					blockTimes = append(blockTimes, blockTime)
+				}
+
+				printStats(start, &producedCount, &sentCount, &prevSentCount, blockHeights, blockTimes)
+				startHeight = currHeight
+				start = time.Now()
+				blockHeights, blockTimes = nil, nil
+
+			case <-done:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+
+	// Wait for a termination signal
+	<-signals
+	fmt.Println("SIGINT received, shutting down...")
+	close(done)
+	wg.Wait()
+	close(txQueue)
+}
+
+func printStats(startTime time.Time, producedCount *int64, sentCount *int64, prevSentCount *int64, blockHeights []int, blockTimes []string) {
+	elapsed := time.Since(startTime)
+	produced := atomic.LoadInt64(producedCount)
+	sent := atomic.LoadInt64(sentCount)
+	tps := float64(sent-*prevSentCount) / elapsed.Seconds()
```

Comment:
b/c sender uses broadcast mode = block, we can use client-side metrics to calculate tps